binary_search_tree/tree.py

```diff
@@ -14,45 +16,148 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        tree = TreeNode(key, value)
+
+        if self.root == None: 
+            self.root = tree
+        else: 
+            current = self.root 
+
+            while current != None: 
+                if key == current.key: 
+                    return None
+                if key < current.key: 
+                    if current.left == None:
+                        current.left = tree
+                    current = current.left 
+                elif key > current.key:
+                    if current.right == None:
+                        current.right = tree
+                    current = current.right 
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        
+        if self.root == None:
+            return None
+
+        current = self.root
+
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def inorder(self):
-        pass
+        tree_nodes = []
+
+        if self.root == None:
+            return tree_nodes
+        else:
+            return self.inorder_helper(self.root, tree_nodes)
+
+    def inorder_helper(self, current, tree_nodes):
+        if current == None:
+            return None
+
+        self.inorder_helper(current.left, tree_nodes)
+        
+        tree_nodes.append(
+                {
+                    "key": current.key, 
+                    "value": current.value
+                }
+            )
 
-    # Time Complexity: 
-    # Space Complexity:     
+        self.inorder_helper(current.right, tree_nodes)
+
+        return tree_nodes
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)     
     def preorder(self):
```

Comment:
👍 However the space complexity is O(n) because you're building a list of all the nodes.